File: Java/090_Subsets2.java
// bacttrack
public class Solution {
	public List<List<Integer>> subsetsWithDup(int[] nums) {
		Arrays.sort(nums);
		List<List<Integer>> res = new ArrayList<List<Integer>>();
		List<Integer> temp = new ArrayList<>();
		subsetsWithDup(res, nums, temp, 0);
	    return res;
	}
	  
	private void subsetsWithDup(List<List<Integer>> res, int[] nums, List<Integer> temp, int begin) {
		List<Integer> sub = new ArrayList<>();
		for (int e : temp) sub.add(e);
		res.add(sub);
		for (int i = begin; i < nums.length; ++i) {
			if (i == begin || nums[i] != nums[i-1]) {
				temp.add(nums[i]);
				subsetsWithDup(res, nums, temp, i+1);
				temp.remove(temp.size()-1);
		    }	  
	    }
	}
}

// iterative
public class Solution {
    public List<List<Integer>> subsetsWithDup(int[] nums) {
    	Arrays.sort(nums);
        List<List<Integer>> res = new ArrayList<>();
        res.add(new ArrayList<>());
        int size = 0, startIndex = 0;
        for (int i = 0; i < nums.length; ++i) {
        	startIndex = i >= 1 && nums[i] == nums[i-1] ? size : 0;
        	size = res.size();
        	for (int j = startIndex; j < size; ++j) {
        		List<Integer> temp = new ArrayList<>();
        		for (int e : res.get(j))
        			temp.add(e);
        		temp.add(nums[i]);
        		res.add(temp);
        	}
        }
        return res;
    }
}

File: Cpp/090_Subsets2.cpp
// backTrack
class Solution {
public:
    vector<vector<int> > subsetsWithDup(vector<int> &nums) {
        sort(nums.begin(), nums.end());
        vector<vector<int>> res;
        vector<int> temp;
        subsetsWithDup(res, nums, temp, 0);
        return res;
    }
    
    void subsetsWithDup(vector<vector<int>>& res, vector<int>& nums, vector<int>& temp, int begin) {
		res.push_back(temp);
		for (int i = begin; i < nums.size(); ++i) {
			if (i == begin || nums[i] != nums[i-1]) {
				temp.push_back(nums[i]);
				subsetsWithDup(res, nums, temp, i+1);
				temp.pop_back();
		    }	  
	    }
	}
};

// iterative
class Solution {
public:
    vector<vector<int> > subsetsWithDup(vector<int> &S) {
        sort(S.begin(), S.end());
        vector<vector<int>> ret = {{}};
        int size = 0, startIndex = 0;
        for (int i = 0; i < S.size(); i++) {
            startIndex = i >= 1 && S[i] == S[i - 1] ? size : 0;
            size = ret.size();
            for (int j = startIndex; j < size; j++) {
                vector<int> temp = ret[j];
                temp.push_back(S[i]);
                ret.push_back(temp);
            }
        }
        return ret;
    }
};
